File: algorithm/src/main/java/baekjoon/p09081/Main.java
package baekjoon.p09081;

import java.util.Arrays;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {
            System.out.println(solve(sc.next()));
        }
    }

    private static String solve(String input) {
        char[] arr = input.toCharArray();
        int cut = cut(arr);
        if (cut == -1) {
            return input;
        }

        int min = min(arr, arr[cut]);
        swap(arr, cut, min);
        Arrays.sort(arr, cut + 1, arr.length);

        return new String(arr);
    }

    private static int cut(char[] arr) {
        for (int i = arr.length - 2; i >= 0; i--) {
            if (arr[i] < arr[i + 1]) {
                return i;
            }
        }

        return -1;
    }

    private static int min(char[] arr, char ch) {
        for (int i = arr.length - 1; i >= 0; i--) {
            if (arr[i] > ch) {
                return i;
            }
        }

        return -1;
    }

    private static void swap(char[] arr, int a, int b) {
        char temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

}

File: algorithm/src/main/java/baekjoon/p09081/Main.kt
package baekjoon.p09081

fun main(args: Array<String>) {
    repeat(readLine()!!.toInt()) {
        println(solve(readLine()!!))
    }
}

private fun solve(input: String): String {
    val arr = input.toCharArray()
    val cut = (arr.lastIndex - 1 downTo 0).find { arr[it] < arr[it + 1] } ?: return input
    val min = (arr.lastIndex downTo cut + 1).find { arr[it] > arr[cut] }!!

    swap(arr, cut, min)
    arr.sort(cut + 1)

    return arr.joinToString("")
}

private fun swap(arr: CharArray, a: Int, b: Int) {
    val aa = arr[a]
    val bb = arr[b]
    arr[a] = bb
    arr[b] = aa
}